mooncake-store/src/pybind_client.cpp

```diff
@@ -79,15 +77,77 @@ void ResourceTracker::signalHandler(int signal) {
 
 void ResourceTracker::exitHandler() { getInstance().cleanupAllResources(); }
 
+void ResourceTracker::startSignalThread() {
+    std::call_once(signal_once_, [this]() {
+        // Wait signal thread start
+        std::promise<void> ready;
+        auto ready_future = ready.get_future();
+
+        // Block signals in this thread; new threads inherit this mask
+        sigset_t set;
+        sigemptyset(&set);
+        sigaddset(&set, SIGINT);
+        sigaddset(&set, SIGTERM);
+        sigaddset(&set, SIGHUP);
+        sigaddset(&set, SIGUSR1);  // used to interrupt sigwait on stop
+        pthread_sigmask(SIG_BLOCK, &set, nullptr);
+
+        signal_thread_ = std::jthread(
+            [set, ready = std::move(ready)](std::stop_token st) mutable {
+                // Register a stop callback to interrupt sigwait via SIGUSR1
+                pthread_t self = pthread_self();
+                std::stop_callback cb(
+                    st, [self]() { pthread_kill(self, SIGUSR1); });
+                ready.set_value();
+                for (;;) {
+                    int sig = 0;
+                    int rc = sigwait(&set, &sig);
+                    if (rc != 0) {
+                        LOG(ERROR) << "sigwait failed: " << strerror(rc);
+                        continue;
+                    }
+
+                    if (sig == SIGUSR1) {
+                        if (st.stop_requested()) {
+                            break;  // graceful stop
+                        }
+                        continue;  // spurious
+                    }
+
+                    // Perform cleanup in normal thread context
+                    LOG(INFO) << "Received signal " << sig
+                              << ", cleaning up resources";
+                    ResourceTracker::getInstance().cleanupAllResources();
+
+                    // Restore default action and re-raise to terminate normally
+                    struct sigaction sa;
+                    sa.sa_handler = SIG_DFL;
+                    sigemptyset(&sa.sa_mask);
+                    sa.sa_flags = 0;
+                    sigaction(sig, &sa, nullptr);
+                    raise(sig);
```

Comment:
Once it finishes cleanup, it'll re-raise the signal to terminate the whole process.

Reply:
There's a really small gap between them. But I think heartbeat may solve this.